dagjose/gen/gen.go

```diff
@@ -0,0 +1,140 @@
+
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/schema"
+	gengo "github.com/ipld/go-ipld-prime/schema/gen/go"
+)
+
+func main() {
+	ts := schema.TypeSystem{}
+	ts.Init()
+
+	// -- Common types -->
+
+	ts.Accumulate(schema.SpawnString("String"))
+	ts.Accumulate(schema.SpawnBytes("Bytes"))
+	ts.Accumulate(schema.SpawnInt("Int"))
+	ts.Accumulate(schema.SpawnFloat("Float"))
+	ts.Accumulate(schema.SpawnLink("Link"))
+	ts.Accumulate(schema.SpawnMap("Map", "String", "Any", false))
+	ts.Accumulate(schema.SpawnList("List", "Any", false))
+
+	// The `Any` union represents a wildcard nested type that can contain any type of scalar or recursive information
+	// including itself (as map values or list elements).
+	ts.Accumulate(schema.SpawnUnion("Any",
+		[]schema.TypeName{
+			"String",
+			"Bytes",
+			"Int",
+			"Float",
+			"Map",
+			"List",
+		},
+		schema.SpawnUnionRepresentationKinded(map[datamodel.Kind]schema.TypeName{
+			datamodel.Kind_String: "String",
+			datamodel.Kind_Bytes:  "Bytes",
+			datamodel.Kind_Int:    "Int",
+			datamodel.Kind_Float:  "Float",
+			datamodel.Kind_Map:    "Map",
+			datamodel.Kind_List:   "List",
+		}),
+	))
+
+	// -- Decode types -->
+
+	// While `Base64Url` is a `String` type and generated through the schema, it has some (surgical) modifications that
+	// allow it to be treated as a base64url-encoded string "lens" looking at raw, un-encoded bytes being decoded.
+	ts.Accumulate(schema.SpawnString("Base64Url"))
+
+	// JWS
+	ts.Accumulate(schema.SpawnStruct("DecodedSignature", []schema.StructField{
+		schema.SpawnStructField("header", "Any", true, false),
+		schema.SpawnStructField("protected", "Base64Url", true, false),
+		schema.SpawnStructField("signature", "Base64Url", false, false),
+	}, schema.SpawnStructRepresentationMap(nil)))
+
+	ts.Accumulate(schema.SpawnList("DecodedSignatures", "DecodedSignature", false))
+
+	// JWE
+	ts.Accumulate(schema.SpawnStruct("DecodedRecipient", []schema.StructField{
+		schema.SpawnStructField("header", "Any", true, false),
+		schema.SpawnStructField("encrypted_key", "Base64Url", true, false),
+	}, schema.SpawnStructRepresentationMap(nil)))
+
+	ts.Accumulate(schema.SpawnList("DecodedRecipients", "DecodedRecipient", false))
+
+	// JOSE
+	ts.Accumulate(schema.SpawnStruct("DecodedJOSE", []schema.StructField{
+		schema.SpawnStructField("aad", "Base64Url", true, false),
+		schema.SpawnStructField("ciphertext", "Base64Url", true, false),
+		schema.SpawnStructField("iv", "Base64Url", true, false),
+		// `link` is not encoded as part of DAG-JOSE because it is not included in the DAG-JOSE spec but is included
+		// here in the schema because it is required when decoding/encoding from/to other encodings (e.g. DAG-JSON).
+		// If `payload` is present during decode, `link` is added with contents matching `payload`. If `link` is present
+		// during encode, it is validated against `payload` and then ignored.
+		schema.SpawnStructField("link", "Link", true, false),
```

Comment:
(just annotating as I go in reading)

This initially seemed strange (and would be, in almost any other use of schemas)... but as long as I keep reminding myself that this isn't really a general purpose schema, it's just meant to used inside one specific codec which is doing special logical operations such as this one described in the comment... then it seems fine.

(And I definitely appreciate the comment being here.)

Reply:
-- wait, sorry.  I don't understand what's going on here at all.

I initially assumed these definitions matched what's in the [dag-jose spec](https://ipld.io/specs/codecs/dag-jose/spec/), but they don't?  There's no `link` fields in this structure as it is stated there?

This is related to my confusion in the other comment in https://github.com/ceramicnetwork/go-dag-jose/pull/23/files?diff=unified&w=0#r749816426 .

Reply:
I'm still wrapping my head around how builders/assemblers can/should be used 😕

I was trying to match the schema from the spec but needed `link` in there to allow "copying" into the `dag-jose` builder from another encoding that included `link`, e.g. `dag-json`. That way, I could accept it into the builder then drop it on the floor during encode.

Reply:
Another thing I could do would be to use the same type of logic I used to "unflatten" nodes (i.e. "flattened" node -> reconstructed map -> "general" node) to skip `link` fields and then remove them from the schema.

Reply:
I think I understand what's up now.  Is "DecodedJOSE" in the code here a combination of the "DecodedJWE" and "DecodedJWS" types as stated in the schema?

I think I understand the desire there (to minimize code?) but this seems relatively risky.  It means we need the "figure out which one a message is, and reject the other fields" in application logic again.  We see some of that with the code around the "link" field, but I sorta suspect that's not all of it, because there's a lot more fields in the "DecodedJWE" struct spec than in the "DecodedJWS" struct spec.

I wonder if this wouldn't be safer and get more work "done for free" by just making both those types fully, and when decoding, just try decoding into each type.  If one fails, try the other; if they both fail, then decode fails.

Reply:
I think what I'd want to test here, in any case, is to make sure we know what fields actually show up when iterating over the data model.  As soon as I can find some time, I'll propose some more fixtures which probe this.  I want to make sure that this implementation and the JS implementation are aligned on that matter, because if not, higher level things like Selectors will start producing nasty surprises.  It's possible that the _combination_ of the dag-jose and the dag-json fixtures in https://github.com/ipld/codec-fixtures/pull/11/ test that, but if so, it's pretty indirectly and a bit hard to call it "legible", so I'd like to add more fixtures which make it clearer.  (And get them exercised from this PR, too, so we're sure.)

Reply:
Hmm, that makes sense. I like the idea of being able to fully rely on the schema to validate the types instead of having an ugly mixture of some schema-based rules and some application ones.

I started out with distinct types but then later combined them to reduce cardinality. It kept things simpler and easier to reason about while prototyping (and that's also how it was done in the prior version of the code, presumably for the same reason lol). I seem to recall some other hurdle I ran into when going with that decision but I've been through so many experiments I can't remember what exactly it was.

Now that I feel much more comfortable with IPLD node assemblers though, I don't think it will be difficult to make and switch to distinct types.

Reply:
`link` will continue to be a special case. We can't encode it (to maintain compatibility with JS encodings) and I also didn't want to not return it from decode so that I could keep behavior identical to JS decode.

I do like your thought of not conflating encode/decode and presentation, so having a separate presentation layer could be clearer and more helpful. That's starting to sound like an ADL lol (IIUC).

Reply:
https://github.com/ipld/ipld/pull/151 are the fixtures I propose to help clarify this.